### DISCUSSION TITLE
How to use UIApplication.shared.open in reducer?

### DISCUSSION THREAD

**Discussion by theMuslimDev** [pointfreeco/swift-composable-architecture#2441] [Q&A] [answered]
I am trying to create a location permission flow. I have a button which shows up if the user has denied location permission. They can click on the button and the action is sent to the reducer. Some checks are made to see if we can show the location permission modal or not. If we can't then using `UIApplication.shared.open(URL(string: UIApplication.openSettingsURLString)!)` I navigate them to the settings page for the app.

However, running `UIApplication.shared.open` directly in the reducer is not right because it will be running in a non-main thread. I can't figure out how I can run `UIApplication.shared.open` directly in the reducer. My initial idea was to use the `@Dependency(\.openUrl)`. This can be used in

```swift

return .run { send in

   await openUrl(URL(string: UIApplication.openSettingsURLString)!)

}
```

This results in a runTImeWarning for TCA 

```swift
ViewStore.send" was called on a non-main thread with...
```

I am able to move `UIApplication.shared.open` into the view and do something like this in the button action. I no longer get the purple warning. However, I would prefer to not have this in the view so it can be testable.. 

```swift
Task { @MainActor in
                                    UIApplication.shared.open(URL(string: UIApplication.openSettingsURLString)!)
 }
```

Any ideas will be helpful. Thanks

**Comment by mbrandonw** [comment, marked as answer]
The `.run` effect you posted above is the correct way to do this, and it cannot cause the runtime warning you showed. That warning only occurs when sending an action directly to a view store, which effects do not do. So there must be something else going on.

If you expand the warning in the left sidebar of Xcode it will show you a stack trace of where the the `viewStore.send` happened that caused the warning. Maybe that will help you track down what happened.

**Reply by theMuslimDev** [reply]
Thanks. Turned out the re-usable button component I was using had the `onButtonTapped` action as `() async -> Void` and the action it self was wrapped in a Task. Something like this.

```swift
Button {
    Task { 
        await onButtonTapped()
    }
 } label: {
            
 }
```

That's why I was getting the runtime warning
